pkg/pillar/zedUpload/datastore_azure.go

```diff
@@ -162,6 +164,31 @@ func (ep *AzureTransportMethod) processAzureUploadByChunks(req *DronaRequest) er
 	return azure.UploadPartByChunk(ep.acName, ep.acKey, ep.container, req.localName, req.UploadID, ep.hClient, req.Adata)
 }
 
+func (ep *AzureTransportMethod) processDownloadByChunks(req *DronaRequest) error {
+	file := req.name
+	dataChan := make(utils.ChunkDataChan)
+	defer close(dataChan)
+	go func(req *DronaRequest, chunkData utils.ChunkDataChan) {
+		for {
+			select {
+			case stats, ok := <-chunkData:
+				if !ok {
+					return
+				}
+				ep.ctx.postChunk(req, stats.Size, stats.Chunk)
+			default:
+				//ignore
+			}
+		}
+	}(req, dataChan)
+	err := azure.DownloadAzureBlobByChunks(ep.acName, ep.acKey, ep.container, file, req.objloc, ep.hClient, dataChan)
+	req.cleanUpByteData()
```

Comment:
How do you wait for the func to complete before this function returns? Can you have the func close dataChan when it is done instead of the defer above?

Reply:
Was that why you had the sleep earlier?

Reply:
If so, shouldn't func also call cleanUpDataBytes before returning?

Reply:
I had added sleep before because s3 download had a issue because it was very frequent write on the channel even before it channel finishes it's job as a result the contents were getting over-written. then changed the downloading logic of s3 for solving the issue.

cleanUpDataBytes is needed only incase of successful download, in case of error scenario we don't care in case, handled as you suggested.
noted, closing the channel when it is done